File: Fuel Tracker by Developer Guidance/[C#]-Fuel Tracker by Developer Guidance/C#/FuelTracker/Model/Fillup.cs
// ===================================================================================
//  Microsoft Developer Guidance
//  Application Guidance for Windows Phone 7 
// ===================================================================================
//  Copyright (c) Microsoft Corporation.  All rights reserved.
//  THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
//  OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
//  LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
//  FITNESS FOR A PARTICULAR PURPOSE.
// ===================================================================================
//  The example companies, organizations, products, domain names,
//  e-mail addresses, logos, people, places, and events depicted
//  herein are fictitious.  No association with any real company,
//  organization, product, domain name, email address, logo, person,
//  places, or events is intended or should be inferred.
// ===================================================================================

using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace FuelTracker.Model
{
    public class Fillup : INotifyPropertyChanged
    {
        private DateTime _date;
        private float _fuelQuantity;
        private float _odometerReading;
        private float _pricePerFuelUnit;

        public DateTime Date
        {
            get { return _date; }
            set
            {
                if (_date == value) return;
                _date = value;
                NotifyPropertyChanged("Date");
            }
        }

        public float OdometerReading
        {
            get { return _odometerReading; }
            set
            {
                var roundedValue = (float)Math.Round(value, 0);
                if (_odometerReading.Equals(roundedValue)) return;
                _odometerReading = roundedValue;
                NotifyPropertyChanged("OdometerReading");
            }
        }

        public float FuelQuantity
        {
            get { return _fuelQuantity; }
            set
            {
                var roundedValue = (float)Math.Round(value, 1);
                if (_fuelQuantity.Equals(roundedValue)) return;
                _fuelQuantity = roundedValue;
                NotifyPropertyChanged("FuelQuantity");
            }
        }

        public float PricePerFuelUnit
        {
            get { return _pricePerFuelUnit; }
            set
            {
                var roundedValue = (float)Math.Round(value, 2);
                if (_pricePerFuelUnit.Equals(roundedValue)) return;
                _pricePerFuelUnit = roundedValue;
                NotifyPropertyChanged("PricePerFuelUnit");
            }
        }

        public float FuelEfficiency
        {
            get { return DistanceDriven / FuelQuantity; }
        }

        public float DistanceDriven { get; set; }

        public IList<string> Validate()
        {
            var validationErrors = new List<string>();

            if (OdometerReading <= 0)
            {
                validationErrors.Add("The odometer value must be a number greater than zero.");
            }

            if (DistanceDriven <= 0)
            { 
                validationErrors.Add("The odometer value must be greater than the previous value.");
            }

            if (FuelQuantity <= 0) 
            {
               validationErrors.Add("The fuel quantity must be greater than zero.");
            }

            if (PricePerFuelUnit <= 0) 
            {
                validationErrors.Add("The fuel price must be greater than zero.");
            }
            
            return validationErrors;
        }

        #region INotifyPropertyChanged

        public event PropertyChangedEventHandler PropertyChanged;

        private void NotifyPropertyChanged(string propertyName)
        {
            var handler = PropertyChanged;
            if (handler != null) handler(this, new PropertyChangedEventArgs(propertyName));
        }

        #endregion
    }
}

File: Fuel Tracker by Developer Guidance/[VB]-Fuel Tracker by Developer Guidance/VB/FuelTracker/Model/Fillup.vb
' ===================================================================================
'  Microsoft Developer Guidance
'  Application Guidance for Windows Phone 7 
' ===================================================================================
'  Copyright (c) Microsoft Corporation.  All rights reserved.
'  THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
'  OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
'  LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
'  FITNESS FOR A PARTICULAR PURPOSE.
' ===================================================================================
'  The example companies, organizations, products, domain names,
'  e-mail addresses, logos, people, places, and events depicted
'  herein are fictitious.  No association with any real company,
'  organization, product, domain name, email address, logo, person,
'  places, or events is intended or should be inferred.
' ===================================================================================

Imports System
Imports System.Collections.Generic
Imports System.ComponentModel

Namespace Model

    Public Class Fillup
        Implements INotifyPropertyChanged

        Private _date As DateTime
        Private _fuelQuantity As Single
        Private _odometerReading As Single
        Private _pricePerFuelUnit As Single

        Public Property [Date] As DateTime
            Get
                Return _date
            End Get
            Set(value As DateTime)
                If _date = value Then Return
                _date = value
                NotifyPropertyChanged("Date")
            End Set
        End Property

        Public Property OdometerReading As Single
            Get
                Return _odometerReading
            End Get
            Set(value As Single)
                Dim roundedValue As Single = Math.Round(value, 0)
                If _odometerReading = roundedValue Then Return
                _odometerReading = roundedValue
                NotifyPropertyChanged("OdometerReading")
            End Set
        End Property

        Public Property FuelQuantity As Single
            Get
                Return _fuelQuantity
            End Get
            Set(value As Single)
                Dim roundedValue As Single = Math.Round(value, 1)
                If _fuelQuantity = roundedValue Then Return
                _fuelQuantity = roundedValue
                NotifyPropertyChanged("FuelQuantity")
            End Set
        End Property

        Public Property PricePerFuelUnit As Single
            Get
                Return _pricePerFuelUnit
            End Get
            Set(value As Single)
                Dim roundedValue As Single = Math.Round(value, 2)
                If _pricePerFuelUnit = roundedValue Then Return
                _pricePerFuelUnit = roundedValue
                NotifyPropertyChanged("PricePerFuelUnit")
            End Set
        End Property

        Public ReadOnly Property FuelEfficiency As Single
            Get
                Return (DistanceDriven / FuelQuantity)
            End Get
        End Property

        Public Property DistanceDriven As Single

        Public Function Validate() As IList(Of String)
            Dim results = New List(Of String)
            If OdometerReading <= 0 Then _
                results.Add("The odometer value must be a number greater than zero.")
            If DistanceDriven <= 0 Then _
                results.Add("The odometer value must be greater than the previous value.")
            If FuelQuantity <= 0 Then _
                results.Add("The fuel quantity must be greater than zero.")
            If PricePerFuelUnit <= 0 Then _
                results.Add("The fuel price must be greater than zero.")
            Return results
        End Function

#Region "INotifyPropertyChanged"

        Public Event PropertyChanged As PropertyChangedEventHandler _
            Implements INotifyPropertyChanged.PropertyChanged

        Private Sub NotifyPropertyChanged(ByVal propertyName As String)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub

#End Region

    End Class

End Namespace
